### PR TITLE
[RFC][Usability] Improve general Ray stacktrace including adding Actor repr

### LINKED ISSUE #17170: Include repr() of the actor in actor task errors
### Describe your feature request

In many cases, IP/hostname/pid are not very interpretable.

Often times, in a user program, actors can have nice interpretable labels (either pre-set or determined during runtime).

Ideally, we can allow the user to do the following:

```
Traceback (most recent call last):
...
ray.exceptions.RayTaskError: ray::...() (label="data_sharder_rank:1")
...
ray.exceptions.RayTaskError: ray::...() (label="trainer_rank:4")
```

and the exception can also programmatically capture this label, like:

```
exc.cause.cause.cause.{label,pid,ip}
```

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Please comment if you don't like the approach here. 

I found out the current Ray stacktrace is ugly, and adding a repr of the actor handle was probably not the only thing I could do. So, I just added more fixes in my mind & add regression tests.

There are 4 different types of stacktrace in Ray.

1. Not nested task (including actor creation) or actor task failure.
2. Chained tasks.
3. Dependency failure (upstreamed dependency raises an exception).

There are important factors.
- The root cause of the failure should be printed at the bottom.
- Ray-related code shouldn't be printed at all to the user-level stacktrace.
- It should be easy to follow stacktrace.

To follow this I completed the below items;
- Skip all stacktraces printed from Ray internal files such as _raylet, client_hook, function manager, and etc.
- Add regression tests.
- Adding a label to the stack trace (which includes repr(actor))
- Add a better human-readable error message for dependency failure. 

Here are examples;
Regular actor task failure
Screen Shot 2021-07-27 at 11 50 54 PM

Chained exceptions from multiple tasks
Screen Shot 2021-07-27 at 11 51 18 PM

**NOTE: This is the previous error message for the same traceback;**
Screen Shot 2021-07-27 at 11 55 20 PM

----------------------------------------------------------------------

Dependency failures
Screen Shot 2021-07-27 at 11 51 31 PM

**NOTE: This is the previous error message for the same traceback**
Screen Shot 2021-07-27 at 11 56 58 PM

## Related issue number

Closes https://github.com/ray-project/ray/issues/17170

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
Hey nice! I like the direction of reducing the "ray specific" stackinfo from the traceback. 

Can you provide the full example script and the traceback that you're getting here?

**Comment by richardliaw** [comment]
In particular, the "task X failed because its input task has failed" seems to be a bit confusing to me (what's an input task?)

I'd instead like to see the stack trace for the user code where a() and f() are being invoked.

**Comment by richardliaw** [comment]
Another comment about the provided example: the default label for the actor is really long and overwhelming. Can we instead just provide the class name (without module namespace)?

File: python/ray/actor.py [hunk 1]
```diff
@@ -949,9 +949,10 @@ def __dir__(self):
         return self._ray_method_signatures.keys()
 
     def __repr__(self):
-        return (f"Actor("
-                f"{self._ray_actor_creation_function_descriptor.class_name},"
-                f"{self._actor_id.hex()})")
+        return ("Actor("
+                "class_name="
+                f"{self._ray_actor_creation_function_descriptor.class_name}, "
+                f"actor_id={self._actor_id.hex()}")
```
Comment (richardliaw) [inline python/ray/actor.py:955]: question: when a user does:

```
@ray.remote
class Actor:
    def __repr__(self):
        return "hi"
```
does this get overriden?

**Comment by rkooo567** [comment]
All the examples are in the test file!

**Comment by rkooo567** [comment]
Let me follow up after eric leaves some comments (so I can address alll at once)

File: python/ray/actor.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/actor.py:955]: It should, but let me test it out tmrw

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -576,13 +576,24 @@ cdef execute_task(
 
             backtrace = ray._private.utils.format_error_message(
                 traceback.format_exc(), task_exception=task_exception)
+
+            # Generate the actor repr from the handle if the failure was
+            # an actor creation task or actor task.
+            actor_handle = None
+            if <int>task_type != <int>TASK_TYPE_NORMAL_TASK:
+                actor_id = core_worker.get_actor_id()
+                actor_handle = core_worker.get_actor_handle(actor_id)
```
Comment (ericl) [inline python/ray/_raylet.pyx:585]: I don't think you want the repr of the actor handle, but the repr of the `actual actor` (self).

File: python/ray/exceptions.py [hunk 3]
```diff
@@ -133,17 +135,37 @@ def __str__(self):
...
+        # which is unnecessary for users.
         in_worker = False
         for line in lines:
             if line.startswith("Traceback "):
                 out.append(f"{colorama.Fore.CYAN}"
                            f"{self.proctitle}"
                            f"{colorama.Fore.RESET} "
-                           f"(pid={self.pid}, ip={self.ip})")
+                           f"(pid={self.pid}, ip={self.ip}"
+                           f", label={self.label})")
```
Comment (ericl) [inline python/ray/exceptions.py:148]: printing ray::bar() here isn't useful, since it's already present in the highlighted part of the exception message. Please remove this for tasks.

File: python/ray/exceptions.py [hunk 4]
```diff
@@ -133,17 +135,37 @@ def __str__(self):
...
             elif in_worker:
                 in_worker = False
-            elif "ray/worker.py" in line or "ray/function_manager.py" in line:
+            elif ("ray/worker.py" in line
+                  or "ray/_private/function_manager.py" in line
+                  or "_private/client_mode_hook.py" in line):
```
Comment (ericl) [inline python/ray/exceptions.py:153]: Nice fix :+1:

Comment (ericl) [reply python/ray/exceptions.py:153]: What if we just considered all _private as internal?

File: python/ray/exceptions.py [hunk 5]
```diff
@@ -133,17 +135,37 @@ def __str__(self):
...
+                # Ray internal code.
+                in_worker = True
+            elif "ray/_raylet.pyx" in line:
                 in_worker = True
+                if "ray._raylet.raise_if_dependency_failed" in line:
+                    # It means the current task is failed
+                    # due to the dependency failure.
+                    # Print out an user-friendly
+                    # message to explain that..
+                    out.append(
```
Comment (ericl) [inline python/ray/exceptions.py:165]: Please remove the yellow color.

File: python/ray/tests/test_traceback.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+If we include unnecessary stacktrace (e.g., logs from internal files),
+these tests will fail.
+"""
+
+
+def test_actor_creation_stacktrace(ray_start_regular):
+    """Test the actor creation task stacktrace.
+
+    Expected output:
+        The actor died because of an error raised in its creation task, ray::A.__init__() (pid=23585, ip=192.168.1.5, label=ray::A.__init__()) # noqa
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:27]:
```suggestion
        The actor died because of an error raised in its creation task, ray::A.__init__() (pid=23585, ip=192.168.1.5) # noqa
```

File: python/ray/tests/test_traceback.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        error_msg = str(ex)
+        error_lines = error_msg.split("\n")
+        len(error_lines) == 6
+
+
+def test_task_stacktrace(ray_start_regular):
+    """Test the normal task stacktrace.
+
+    Expected output:
+        ray::f() (pid=23839, ip=192.168.1.5, label=ray::f()) # noqa
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:60]:
```suggestion
        ray::f() (pid=23839, ip=192.168.1.5) # noqa
```

File: python/ray/tests/test_traceback.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        error_msg = str(ex)
+        error_lines = error_msg.split("\n")
+        len(error_lines) == 6
+
+
+def test_actor_task_stacktrace(ray_start_regular):
+    """Test the actor task stacktrace.
+
+    Expected output:
+        ray::A.f() (pid=24029, ip=192.168.1.5, label=Actor(class_name=test_actor_task_stacktrace.<locals>.A, actor_id=99da335726d2f7fa26e5c83201000000) # noqa
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:92]:
```suggestion
        ray::A.f() (pid=24029, ip=192.168.1.5, label=Actor(repr=MyActorRepr(), id=99da335726d2f7fa26e5c83201000000) # noqa
```

File: python/ray/tests/test_traceback.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        error_msg = str(ex)
+        error_lines = error_msg.split("\n")
+        len(error_lines) == 6
+
+
+def test_exception_chain(ray_start_regular):
+    """Test the chained stacktrace.
+    
+    Expected output:
+        ray::foo() (pid=24171, ip=192.168.1.5, label=ray::foo()) # noqa
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:125]:
```suggestion
        ray::foo() (pid=24171, ip=192.168.1.5) # noqa
```

File: python/ray/tests/test_traceback.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 127, in foo
+            return ray.get(bar.remote())
+        ray.exceptions.RayTaskError(ZeroDivisionError): ray::bar() (pid=24186, ip=192.168.1.5, label=ray::bar())
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:128]:
```suggestion
        ray.exceptions.RayTaskError(ZeroDivisionError): ray::bar() (pid=24186, ip=192.168.1.5)
```

File: python/ray/tests/test_traceback.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        error_msg = str(ex)
+        error_lines = error_msg.split("\n")
+        len(error_lines) == 7
+
+
+def test_dep_failure(ray_start_regular):
+    """Test the stacktrace genereated due to dependency failures.
+    
+    Expected output:
+        ray::f() (pid=24413, ip=192.168.1.5, label=ray::f()) # noqa
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:157]:
```suggestion
        ray::f() (pid=24413, ip=192.168.1.5) # noqa
```

File: python/ray/tests/test_traceback.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+          The task, ray::f(), failed because the below input task has failed.
+        ray.exceptions.RayTaskError: ray::a() (pid=24413, ip=192.168.1.5, label=ray::a())
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:159]:
```suggestion
        ray.exceptions.RayTaskError: ray::a() (pid=24413, ip=192.168.1.5)
```

File: python/ray/tests/test_traceback.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+          The task, ray::a(), failed because the below input task has failed.
+        ray.exceptions.RayTaskError: ray::b() (pid=24413, ip=192.168.1.5, label=ray::b())
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:161]:
```suggestion
        ray.exceptions.RayTaskError: ray::b() (pid=24413, ip=192.168.1.5)
```

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/exceptions.py (hunk 4 above)
Comment (rkooo567) [reply python/ray/exceptions.py:153]: Sounds good. Also added tracing_helper.py

File: python/ray/_raylet.pyx (hunk 2 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:585]: actual actor means the class itself, am I correct?

Comment (rkooo567) [reply python/ray/_raylet.pyx:585]: Or you are talking about `ActorClass`?

Comment (rkooo567) [reply python/ray/_raylet.pyx:585]: It looks like in this code, it is a bit tricky to get ActorClass, but we can easily get the original class. But in that case, repr is like this by default; `test_traceback.A object at 0x7fa0653b8750>`

Comment (rkooo567) [reply python/ray/_raylet.pyx:585]: So, it looks like we can only obtain the original class here, but if so, the default looks like this`test_traceback.A object at 0x7fb517805710>`. But if users override their repr, that will be reflected. Does this look okay to you?

File: python/ray/actor.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/actor.py:955]: This should work now. I added a test as well. One caveat is the default is the default repr

File: python/ray/_raylet.pyx (hunk 2 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:585]: Actually, this is the mechanism I am trying. Please lmk if you have other good ideas;

If the actor class repr is default => repr of actor handle
otherwise => actor class' repr

Comment (ericl) [reply python/ray/_raylet.pyx:585]: The class itself, `self`, which is the only one containing useful information.

> istest_traceback.A object at 0x7fb517805710>. But if users override their repr, that will be reflected. Does this look okay to you?

It's fine. You could also report `repr=None` but that might be confusing.

Comment (rkooo567) [reply python/ray/_raylet.pyx:585]: Here's the current implementation; 

If actor class doesn't specify repr (in this case, it will be like `test_traceback.A object at 0x7fb517805710>`); We capture it and format it in this way;
```
ray::A.f() (pid=70926, ip=192.168.1.5, repr=Actor(repr=A))
```

If actor class does (e.g., repr returns "ABC");
```
ray::A.f() (pid=70932, ip=192.168.1.5, repr=ABC)
```

Comment (rkooo567) [reply python/ray/_raylet.pyx]: We can capture this by checking repr(actor) == object.__repr__(actor)

Comment (ericl) [reply python/ray/_raylet.pyx]: This is weird. Can you either set it to None or use repr verbatim?

Comment (rkooo567) [reply python/ray/_raylet.pyx]: Sure. I will then just print the original repr in this case

**Comment by ericl** [comment]
Btw do we also include the name of the actor (set with .options)?

**Comment by rkooo567** [comment]
I tried, but it seemed like we needed some refactoring to make that work (so I quickly gave up). Let me check one more time.

**Comment by ericl** [comment]
Is it part of the runtime context?

File: python/ray/_private/function_manager.py [hunk 14]
```diff
@@ -31,6 +31,7 @@
     is_static_method,
 )
 
+# TODO(sang): Include function arg name to function name?
```
Comment (ericl) [inline python/ray/_private/function_manager.py:34]: Remove

File: python/ray/_raylet.pyx [hunk 15]
```diff
@@ -576,13 +577,19 @@ cdef execute_task(
 
             backtrace = ray._private.utils.format_error_message(
                 traceback.format_exc(), task_exception=task_exception)
+
+            # Generate the actor repr from the actor class.
+            error_label = repr(actor) if actor else None
```
Comment (ericl) [inline python/ray/_raylet.pyx:582]:
```suggestion
            actor_repr = repr(actor) if actor else None
```

File: python/ray/_raylet.pyx [hunk 16]
```diff
@@ -576,13 +577,19 @@ cdef execute_task(
...
+
             if isinstance(error, RayTaskError):
                 # Avoid recursive nesting of RayTaskError.
                 failure_object = RayTaskError(function_name, backtrace,
-                                              error.cause, proctitle=title)
+                                              error.cause, proctitle=title,
+                                              label=error_label)
```
Comment (ericl) [inline python/ray/_raylet.pyx:588]:
```suggestion
                                              actor_repr=error_label)
```

File: python/ray/actor.py [hunk 17]
```diff
@@ -949,9 +949,10 @@ def __dir__(self):
         return self._ray_method_signatures.keys()
 
     def __repr__(self):
-        return (f"Actor("
-                f"{self._ray_actor_creation_function_descriptor.class_name},"
-                f"{self._actor_id.hex()})")
+        return ("Actor("
```
Comment (ericl) [inline python/ray/actor.py:952]: Can you revert this change? Unrelated to the PR.

File: python/ray/exceptions.py [hunk 18]
```diff
@@ -72,7 +72,8 @@ def __init__(self,
                  cause,
                  proctitle=None,
                  pid=None,
-                 ip=None):
+                 ip=None,
+                 label=None):
```
Comment (ericl) [inline python/ray/exceptions.py:76]:
```suggestion
                 actor_repr=None):
```

File: python/ray/exceptions.py [hunk 19]
```diff
@@ -133,17 +135,39 @@ def __str__(self):
...
-            elif "ray/worker.py" in line or "ray/function_manager.py" in line:
+            elif ("ray/worker.py" in line or "ray/_private/" in line
+                  or "ray/util/tracing/tracing_helper.py" in line):
+                # Skip internal files that could be printed.
+                # Otherwise, users will see unnecessary stacktrace from
+                # Ray internal code.
+                in_worker = True
+            elif "ray/_raylet.pyx" in line:
                 in_worker = True
+                if "ray._raylet.raise_if_dependency_failed" in line:
```
Comment (ericl) [inline python/ray/exceptions.py:163]: Let's remove for now

File: python/ray/tests/test_traceback.py [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+        error_lines = error_msg.split("\n")
+        len(error_lines) == 7
+
+
+def test_dep_failure(ray_start_regular):
+    """Test the stacktrace genereated due to dependency failures.
+    
+    Expected output:
+        ray::f() (pid=24413, ip=192.168.1.5) # noqa
+          The task, ray::f(), failed because the below input task has failed.
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:158]:
```suggestion
          Some of the input arguments for this task could not be computed:
```

File: python/ray/exceptions.py (hunk 19 above)
Comment (rkooo567) [reply python/ray/exceptions.py:163]: I just changed wording instead.

File: python/ray/tests/test_traceback.py (hunk 9 above)
Comment (ericl) [reply python/ray/tests/test_traceback.py:125]: Not fixed yet.

File: python/ray/tests/test_traceback.py [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+        error_msg = str(ex)
+        error_lines = error_msg.split("\n")
+        assert len(error_lines) == 7
+
+
+def test_dep_failure(ray_start_regular):
+    """Test the stacktrace genereated due to dependency failures.
+    
+    Expected output:
+        ray::f() (pid=24413, ip=192.168.1.5) # noqa
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:157]: Needs update.

**Review by ericl — APPROVED** [review]
LGTM, but for the tests can we actually test the exact stack output? You can do that by scrubbing out the numbers/random parts of the stack trace so it becomes deterministic, like

```
    Expected output:
        ray::foo() (pid=ZZZ, ip=ZZZ, label=ray::foo()) # noqa
        File "STRING", line ZZZ, in foo
            return ray.get(bar.remote())
        ray.exceptions.RayTaskError(ZeroDivisionError): ray::bar() (pid=ZZZ, ip=ZZZ)
        File "STRING", line ZZZ, in bar
            return 1 / 0
        ZeroDivisionError: division by zero
```

**Comment by rkooo567** [comment]
SGTM. I will update tests after fixing the existing issues in the CI

**Comment by rkooo567** [comment]
TODO: Need to compare the exact stacktrace.

**Comment by rkooo567** [comment]
@ericl This is pretty complicated to compare stacktrace... do you think it is worth efforts? https://github.com/ray-project/ray/pull/17389/commits/8c836c3b29d3cd6e2084bf78a244e809704dc8e6

Feel like comparing lines could give us good enough signals, but if you'd like me to do this, I can just try for every test.

**Comment by ericl** [comment]
You can write a generic regex that can be used to scrub all the examples.
Then you just need to check:

Sanitize(actual) == Sanitize(expected).

**Comment by ericl** [comment]
Regex replace, if it wasn't clear.

**Comment by rkooo567** [comment]
@ericl Addressed this; https://github.com/ray-project/ray/pull/17389#issuecomment-891608750

Thanks for the good suggestion btw!! It worked very well. (and clean)

**Comment by rkooo567** [comment]
Btw I skipped Windows tests because in Windows hiding internal file name was never working anyway (the file seperator seems to be \ instead of /).

File: python/ray/tests/test_traceback.py [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+    sys.platform == "win32",
+    reason="Clean stacktrace not supported on Windows")
+def test_task_stacktrace(ray_start_regular):
+    """Test the normal task stacktrace."""
+    expected_output = """ray::f() (pid=57290, ip=192.168.1.5) # noqa
+  File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 96, in f
+    return g(c)
+  File "/Users/sangbincho/work/ray/python/ray/tests/test_traceback.py", line 88, in g
+    raise ValueError(a)
+ValueError: 7"""
```
Comment (ericl) [inline python/ray/tests/test_traceback.py:85]: Should probably commit post scrubbed output here, to avoid regex accidentally replacing everything bugs.

Comment (rkooo567) [reply python/ray/tests/test_traceback.py:85]: Oh yeah good point. I will modify that

**Comment by rkooo567** [comment]
All failing tests are flaky in the master.
